Add common API to emit events

clobrano opened the pull request:
Add common API to emit Events to share the same format among operators.

clobrano committed:
Add API to emit Events with shared format

Signed-off-by: Carlo Lobrano <c.lobrano@gmail.com>

clobrano committed:
Update go.mod

Signed-off-by: Carlo Lobrano <c.lobrano@gmail.com>

slintes commented on pkg/events/events.go:
```diff
@@ -0,0 +1,34 @@
+)
+
+// Event message format "medik8s <operator shortname> <message>"
+const format = "medik8s %3s %s"
```
I think I would skip "medik8s", it's not that helpful, especially downsytream. And I would somehow separate operator and message. Maybe `"%3s: %s"`? Or `"[%3s] %s"`?

clobrano replied:
The idea of "medik8s" string was to help filtering our events

> Maybe "%3s: %s"? Or "[%3s] %s"?

agreed on this
EDIT, I like the latter more

slintes commented on pkg/events/events.go:
```diff
@@ -0,0 +1,34 @@
+
+type EventRecorder struct {
+	record.EventRecorder
+	operator string
```
I think this needs to be public, otherwise you can´t set it when creating this struct in another package?

mshitrit replied:
Nit: IMO might be even nicer to use a method to initialize it instead

clobrano replied:
Agree on both :)

mshitrit commented on pkg/events/events.go:
```diff
@@ -0,0 +1,34 @@
+// Event message format "medik8s <operator shortname> <message>"
+const format = "medik8s %3s %s"
+
+type EventRecorder struct {
```
IIUC this struct implements the interface of `record.EventRecorder`
it also adds 4 additional methods (NormalEvent,  WarningEvent , etc... )

- I think it should also override the original methods of `record.EventRecorder` with our formatted message.
- Nit: I'd consider returning an interface instead of a struct

mshitrit commented on pkg/events/events.go:
```diff
@@ -0,0 +1,77 @@
+	operator string
+}
+
+var _ EventRecorder = &eventRecorder{}
```
hmm I missing something, why is this needed  ? 🤔

clobrano replied:
Not needed in fact 🫣

mshitrit commented on pkg/events/events.go:
```diff
@@ -0,0 +1,77 @@
+	e.WarningEvent(object, reason, msg)
+}
+
+func (e *eventRecorder) Event(object runtime.Object, eventtype, reason, message string) {
```
replace `eventtype` with `eventType`

clobrano replied:
I copy/pasted the signature from upstream, but I can change it :)

https://pkg.go.dev/k8s.io/client-go/tools/record#EventRecorder

mshitrit commented on pkg/events/events.go:
```diff
@@ -0,0 +1,77 @@
+)
+
+// Event message format "medik8s <operator shortname> <message>"
+const customFmt = "[%3s] %s"
```
I am missing the key word ("medik8s", "remediation" or whatever we agree on) that would help searching our events.
IIUC it should be added here instead of sent by each operator.

clobrano replied:
Drop it by mistake, thanks for noticing.
At the moment we didn't completely agreed on medik8s or remediation though :shrug:

mshitrit commented on pkg/events/events.go:
```diff
@@ -0,0 +1,77 @@
+	}
+}
+
+func (e *eventRecorder) NormalEvent(object runtime.Object, reason, message string) {
+	e.Event(object, "Normal", reason, message)
+}
+
+func (e *eventRecorder) NormalEventf(object runtime.Object, reason, messageFmt string, a ...interface{}) {
+	msg := fmt.Sprintf(messageFmt, a...)
+	e.NormalEvent(object, reason, msg)
+}
+
+func (e *eventRecorder) WarningEvent(object runtime.Object, reason, message string) {
+	e.Event(object, "Warning", reason, message)
+}
+
+func (e *eventRecorder) WarningEventf(object runtime.Object, reason, messageFmt string, a ...interface{}) {
+	msg := fmt.Sprintf(messageFmt, a...)
+	e.WarningEvent(object, reason, msg)
+}
```
Why not add the formatted message with keyword and operator name to these 4 methods ? 🤔

clobrano replied:
I changed the code, but for completeness :grin: I was adding the formatted message here

`e.WarningEvent` was calling the customized `Event` function which added the shared format

mshitrit commented on pkg/events/events_test.go:
```diff
@@ -0,0 +1,64 @@
+			Entry("Emit normal event with args",
+				nil, "Normal", "thisReason", "normal event message with %d %s",
+				"Normal thisReason [remediation] normal event message with 2 args",
+				2, "args"),
```
Nit: IIUC this event message has two arguments: `2` and "args" I think using "agrs" as an argument is a bit confusing

mshitrit commented on pkg/events/events_test.go:
```diff
@@ -0,0 +1,64 @@
+			Entry("Emit warning event with args",
+				nil, "Warning", "thisReason", "warning event message with %d %s",
+				"Warning thisReason [remediation] warning event message with 4 args",
+				4, "args"),
```
I think this test is confusing, I'm expecting four arguments for the event message but in fact there are only two: `4` and `args`

razo7 commented on pkg/events/events.go:
```diff
@@ -0,0 +1,50 @@
+
+// RemediationStoppedByNHC will record a Normal event with reason RemediationStopped.
+func RemediationStoppedByNHC(recorder record.EventRecorder, object runtime.Object) {
+	NormalEvent(recorder, object, "RemediationStopped", "NHC added the timed-out annotation, remediation will be stopped")
```
Is it possible to save the event's reason and message in a const which will be used here and in [`verifyEvent`](https://github.com/medik8s/machine-deletion-remediation/pull/96/files#diff-867a8834813cdf408e2387961fc75a31146379802d5abec6f465f1d8a565b903R477) function for unit tests?

clobrano replied:
I'd rather keep it this way. I understand it creates a bit of frictions writing the tests, but I think it's a good friction. It would improve test readability (you actually read in the test source code what text is expected, without looking at the variable's value) and would spot typos easily.

razo7 replied:
I have no strong preference, fine by me.

razo7 commented:
/lgtm
Giving a chance to get more feedbacks, feel free to unhold
/hold

slintes commented on pkg/events/events.go:
```diff
@@ -0,0 +1,50 @@
+// Event message format "medik8s <operator shortname> <message>"
+const customFmt = "[remediation] %s"
+
+// NormalEventf will record an event with type Normal and fixed message.
```
typo: NormalEvent*f*

slintes commented on pkg/events/events.go:
```diff
@@ -0,0 +1,50 @@
+)
+
+// Event message format "medik8s <operator shortname> <message>"
+const customFmt = "[remediation] %s"
```
any reason to not use this? 🤔

clobrano replied:
none indeed, it was intended to be used :shrug: 
I'll fix this

slintes commented on pkg/events/events.go:
```diff
@@ -0,0 +1,50 @@
+
+// NormalEventf will record an event with type Normal and fixed message.
+func NormalEvent(recorder record.EventRecorder, object runtime.Object, reason, message string) {
+	recorder.Event(object, "Normal", reason, fmt.Sprintf("[remediation] %s", message))
```
please use corev1.EventTypeNormal
(similar at other places)

slintes commented on pkg/events/events.go:
```diff
@@ -0,0 +1,50 @@
+}
+
+// WarningEventf will record an event with type Warning and fixed message.
+func WarningEvent(recorder record.EventRecorder, object runtime.Object, reason, message string) {
```
typo: WarningEvent*f*

clobrano committed:
fixup! Add API to emit Events with shared format

Actually use customFmt var, event type definitions from
k8s.io/api/core/v1 and fix typos

slintes commented:
/lgtm
